compiler/rustc_parse/src/parser/stmt.rs

```diff
@@ -155,17 +155,20 @@ impl<'a> Parser<'a> {
 
         let mac = MacCall { path, args, prior_type_ascription: self.last_type_ascription };
 
-        let kind = if delim == token::Brace || self.token == token::Semi || self.token == token::Eof
-        {
-            StmtKind::MacCall(P(MacCallStmt { mac, style, attrs, tokens: None }))
-        } else {
-            // Since none of the above applied, this is an expression statement macro.
-            let e = self.mk_expr(lo.to(hi), ExprKind::MacCall(mac), AttrVec::new());
-            let e = self.maybe_recover_from_bad_qpath(e, true)?;
-            let e = self.parse_dot_or_call_expr_with(e, lo, attrs.into())?;
-            let e = self.parse_assoc_expr_with(0, LhsExpr::AlreadyParsed(e))?;
-            StmtKind::Expr(e)
-        };
+        let kind =
+            if (delim == token::Brace && self.token != token::Dot && self.token != token::Question)
```

Comment:
Could this condition be split out into a variable?

Reply:
Side observation: the reference's specification of the grammar is fairly inaccurate concerning this specific behaviour. I was going to check if the reference needs adjusting for this change, but it seems like the grammar already allows treating `m!{ ... }` exactly like `m![...]` and `m!(...)` in any `ExpressionWithoutBlock` position.

If we go by reference the `delim == token::Brace` check should not be present here at all.

In practice I suspect that we want to make `m! { ... }` be considered a `ExpressionWithBlock` and allow use of `m! { ... }` anywhere where other `ExpressionWithBlock`s are usable. One example that comes to mind immediately is `vec! { ... }[0]` or some such.

Reply:
That would be a breaking change.

This is already accepted as two statements:

```rust
bla! {}
[0];
```

Reply:
Aha, right, my bad. Upon further checking [this also holds true for other `ExpressionWithBlock`s](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=18966072e2da8e8a0cd239483fb4e2a1) (in that `ExpressionWithBlock [idx]` parses as two statements)  though, apparently, so the point still stands I feel.

Reply:
There are only two cases where an ExpressionWithBlock continues after the `}`: When followed by either a `.` or a `?`. So this PR makes that consistent with `m!{}`.

Reply:
My understanding of the description was exactly as @m-ou-se described, that this PR made `Foo { }` and `m!{ }` consistent. I agree that they should be consistent.